src/DotNetWorker.Core/Diagnostics/FunctionActivitySource.cs

```diff
@@ -0,0 +1,64 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.Azure.Functions.Worker.Diagnostics
+{
+    internal class FunctionActivitySourceFactory
+    {
+        private static readonly ActivitySource _activitySource = new(TraceConstants.FunctionsActivitySource, TraceConstants.FunctionsActivitySourceVersion);
+        private readonly string _schemaVersionUrl;
+        private readonly Lazy<IReadOnlyDictionary<string, string>> _attributeMap;
+
+        public FunctionActivitySourceFactory(IOptions<WorkerOptions> options)
+        {
+            _attributeMap = new Lazy<IReadOnlyDictionary<string, string>>(() => GetMapping(options.Value.OpenTelemetrySchemaVersion));
+            _schemaVersionUrl = TraceConstants.OpenTelemetrySchemaMap[options.Value.OpenTelemetrySchemaVersion];
+        }
+
+        public Activity? StartInvoke(FunctionContext context)
+        {
+            var activity = _activitySource.StartActivity(TraceConstants.FunctionsInvokeActivityName, ActivityKind.Internal, context.TraceContext.TraceParent,
```

Comment:
Should this be `ActivityKind.Server` as we are ultimately serving a function invocation request?

Reply:
See the comment here -- https://github.com/Azure/azure-functions-dotnet-worker/pull/1307#discussion_r1095296512. It seems that this is currently not really defined by the otel spec, but that App Insights and others treat this as an internal call with the host log being the "Server" piece. If they were both "Server", rendering of the Span would not be correct.

I think this is something we can handle with the OTel schema versioning we're doing -- if things change in the future, we can use that to change the Kind as well.

Reply:
I think we need to have a discussion about the spans we will all have. I don't think we should be trying to hide the gRPC hop from Host to Worker. Our Host/Worker design is a distributed system and it should be represented as such in spans. Customers should see the following at minimum:

Incoming request (Host Server Span) --> Host Calling Worker (Host Client Span) --> Worker handling invocation (Worker Server Span)

And then we also have to see how gRPC spans factor into all of this. Given that N gRPC spans may relate to a single invocation, I think they should be links on the Functions spans and not parent/child

Reply:
So -- the spec does say it should be Server: https://opentelemetry.io/docs/reference/specification/trace/semantic_conventions/faas/#incoming-invocations. But I also see really no notion of a "host". It seems to only care about the instance handling the invocation itself.

I think in most other clouds, the "host" -- the thing polling/sending you events -- is part of infrastructure and not even present in the distributed trace? anyone have any idea?

@lmolkova -- if we do indeed make this a Server (which is what it seems like it should be) -- that would turn this span into a Request in App Insights, right?

Reply:
Yes, `Server` becomes `request` in application insights. This is also beneficial because app insights has built in aggregation for request telemetry - failure rate, performance, etc. I see this as a great benefit for customers.

Reply:
If host reports queue trigger, it will report it as consumer, but I think the question here is how should we treat worker, i.e.
Option 1:
- Function-host (Server/Consumer span)
  - Call to worker (client span)
    - Incoming on worker (Server span)

Option 2:
- Function-host (Server/Consumer span)
  - Incoming on worker (Internal span)


Option 3 (possible with hacks only):
- Function-host (Server/Consumer span)
  - no new span for worker invocation, but host context is picked up in `Activity.Current`

I.e. if we consider host and worker as the same instance and want to hide that they are different processed, we should pick option 2. Otherwise we should proceed with option 1. 

I agree that FaaS otel spec does not specify it at all, but should at least recommend a default. 

My opinion here is that Option2 seems like a good start:
- presumably, communication between worker and host is super-fast and reliable and therefore is not quite interesting. Users might not like the idea to pay for extra telemetry that does not explain anything new. I would even prefer option 3 if only it was possible without hacks
- it keeps the door open to implement Option1 or 3 in the future with minimal behavior changes to users

Reply:
Created https://github.com/open-telemetry/opentelemetry-specification/issues/3213

Reply:
Ah, yes. Any other ActivitySource will be differentiated. That's why this one is named with ".Worker" -- it's only meant to emit Activities directly related to the Worker.

Any Activity related to the host will be appropriately named and, won't exist in the worker process -- those will be emitted via the Host process.

Does that address the concern?

Reply:
> My justification is I believe that we should give Server spans for the Worker (and specifically not Internal)

Having two nested server spans will break the application map in application insights and could be confusing to other backends. It also goes against otel spec that says server spans should have client parents - this creates expectations on backends that would be broken. So if the decision is to use server span, it'd be best to add client span that tracks call to worker right away

Reply:
Yes, option 1 does include the client span from Host -> Worker.